[#68031] Fix toggleable fieldsets

myabc opened the pull request:
# Ticket

https://community.openproject.org/wp/68031

# What are you trying to accomplish?

Ensure toggleable fieldsets have their event listeners attached exactly once on turbo:load to avoid duplicate bindings and broken toggling when navigating with Turbo.

## Screenshots
### Reporting

| Open |  Closed |
|--------|--------|
| <img width="1320" height="1193" alt="Screenshot 2025-10-23 at 14 45 50" src="https://github.com/user-attachments/assets/e8ab2c75-f838-46eb-b519-4ad2eddb54d8" /> | <img width="1335" height="959" alt="Screenshot 2025-10-23 at 14 45 27" src="https://github.com/user-attachments/assets/73676d89-4f58-4106-ab60-63d898671067" /> | 

### Versions

<img width="707" height="598" alt="Screenshot 2025-10-23 at 14 06 38" src="https://github.com/user-attachments/assets/50945106-403c-49a4-8f23-7680479c2527" />

### Admin - Repositories

<img width="1319" height="455" alt="Screenshot 2025-10-23 at 14 33 47" src="https://github.com/user-attachments/assets/49ef42e2-877c-4cbf-815a-cf7360558780" />

### Budgets

<img width="1220" height="627" alt="image" src="https://github.com/user-attachments/assets/a06c57d2-0c9b-4470-8e81-4ee273713ab8" />

# What approach did you choose and why?

- Re-attaches the initialization of toggleable fieldsets to the `turbo:load` event so listeners are only bound once.
- Partially reverts changes from #19495 which caused listeners to be attached multiple times or removed incorrectly.
- Addresses the bug where fieldsets did not toggle after Turbo navigations.
- Thus PR also Primerizes some buttons in the Budgets subform - fixing ERB Lint accessibility errors.


# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)
- [x] Manual testing with Turbo navigation recommended across major browsers.

myabc committed:
[#68031] Fix duplicate global listeners

Ensure event listeners are only attached once - on either full page load
(`DOMContentLoaded`) or Turbo page renders (`turbo:render`).

HDinger reviewed (commented):
I am not too sure about this change.. Yes, it fixes the issue but looking at the commit message of the deleted line, it was written:

> [#65298] Fix broken top bar menus on form errors
>
> Menus in the top bar were not properly initialised after unsuccessful
form submissions that result in an 422 Unprocessable Entity response
via Turbo Drive.
>
>This commit updates Angular-Turbo Drive integration code to listen to
`turbo:render` in addition to `turbo:load`. `turbo:load` is only fired
on Turbo Drive _visits_. However, a form submission response that is not
a 303 Redirect is not considered a _visit_ by Turbo Drive.
> 
> See: https://github.com/hotwired/turbo/issues/85
> See also: https://turbo.hotwired.dev/handbook/drive#redirecting-after-a-form-submission


So I guess, this might introduce a regression?

Further, intependent of that: How does tubro interfere with the toggling mechanism at all? And did you try to replace that with the `CollapsibleSection` component we have in Primer?

myabc committed:
Replace toggleable fieldsets: Reporting widgets

Replaces togglable fieldsets with `CollapsibleSection`.

myabc committed:
Replace toggleable fieldsets: Versions index

Replaces togglable fieldsets with `CollapsibleSection`.

myabc committed:
Replace toggleable fieldsets: Budgets subforms

Replaces togglable fieldsets with `CollapsibleSection`.

myabc committed:
Remove legacy toggleable fieldset behavior, styles

myabc committed:
Replace toggleable fieldsets: Admin > Roles

Replaces togglable fieldsets with `CollapsibleSection`.

myabc committed:
Replace toggleable fieldsets: Admin > Repositories

Replaces togglable fieldsets with `CollapsibleSection`.

myabc commented:
> and did you try to replace that with the CollapsibleSection component we have in Primer?

@HDinger I've gone ahead and done that.

myabc commented:
@copilot can you update the PR description for me to reflect the newest changes?

HDinger commented on app/views/roles/_permissions.html.erb:
```diff
@@ -28,29 +28,29 @@ See COPYRIGHT and LICENSE files for more details.
-      </p>
-    </div>
     <% end %>
-    <div class="-columns-2">
```
The two column layout was indeed intended, so I guess we should not simply remove it

HDinger commented on app/views/admin/settings/repositories_settings/_repositories_checkout.html.erb:
```diff
@@ -4,38 +4,37 @@
-          <div class="form--field-container">
-            <%= styled_check_box_tag key, 1, enabled, class: "settings-repositories--checkout-toggle" %>
+      <%= render Primer::OpenProject::CollapsibleSection.new(collapsed: true, display: :block, mb: 3) do |section| %>
+        <% section.with_title(tag: :h5) { klass.vendor_name } %>
```
Why `h5`? There is only a `H2` on that page and a fieldset which could be treated as a `H3`. So that would be a `h4`

myabc replied:
@HDinger I think I was trying to make it visually smaller, but I agree the semantics are incorrect.

myabc replied:
@HDinger Changed to an `h4`. However I've applied a smaller font size which I think looks better inside the existing fieldset.

HDinger commented on frontend/src/app/core/setup/globals/global-listeners/toggable-fieldset.ts:
Nice 👍

HDinger reviewed (changes requested):
Nice job 👏 I have some minor remarks in code. In general, I wonder wether we should att the option to show a border to the `CollapsibleSection`. Many of the view depend on the visual separation by that line at the pages now look weird without it. What do you think? Might also be a topic for tomorrows UX daily

myabc committed:
Fix excessive top margin on Reporting page filters

Only applies a top margin when filter list is used within an
`advanced-filters--container` (e.g. Projects page).

myabc commented:
@HDinger thanks for the feedback. I have gone through each page and checked the heading hierarchies: the semantics should now be correct. Additionally, I've tweaked margins and set `font_size` on section titles to appropriate values: the default [always applies `.f3`](https://primer.style/product/css-utilities/typography/#type-scale-utilities), no matter the header tag (this is something we might want to address upstream?).

Screenshots in the description have also been updated.

myabc commented on frontend/src/turbo/turbo-global-listeners.ts:
```diff
@@ -65,7 +61,7 @@ export function addTurboGlobalListeners() {
   };
   document.addEventListener('turbo:render', runOnRenderAndLoad);
-  document.addEventListener('turbo:load', runOnRenderAndLoad);
+  document.addEventListener('DOMContentLoaded', runOnRenderAndLoad);
```
actually, `turbo:render` and `DOMContentLoaded` should cover all bases. / @oliverguenther

myabc committed:
Fix A11y errors: Primerize Budget subform buttons

Fixes the following ERB Lint errors:

    GitHub::Accessibility::LinkHasHref:Links should go somewhere, you probably want to use a `<button>` instead.
    Learn more at https://github.com/github/erblint-github#rules.

    <a class="budget-add-row wp-inline-create--add-link"
              role="button"
              title="<%= t(:button_add_budget_item) %>"
              data-action="costs--budget-subform#addRow"> (not autocorrected)
    In file: modules/budgets/app/views/budgets/subform/_material_budget_subform.html.erb:124

    GitHub::Accessibility::NoTitleAttribute:The title attribute should never be used as it is inaccessible for several groups of users. Exceptions are provided for <iframe> and <link>.
    Learn more at https://github.com/github/erblint-github#rules.

    <a class="budget-add-row wp-inline-create--add-link"
              role="button"
              title="<%= t(:button_add_budget_item) %>"
              data-action="costs--budget-subform#addRow"> (not autocorrected)
    In file: modules/budgets/app/views/budgets/subform/_material_budget_subform.html.erb:124
